### PR TITLE
Stage1

### PR DESCRIPTION
**Pull request by Xuezhi-Liang**
Unknown

Notes: 

1. The first step is to create the Context class and start to move the core AdaptDL mathematics into that class. 

2. The AdaptiveDataLoader should get refactored to use the Context class for those AdaptDL logic to prevent duplication of logic. 

3. But the changes to AdaptiveDataLoader are purely implementation changes, users of the class shouldn't need to see any difference.

### CODE REVIEW

File: adaptdl/adaptdl/torch/context.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+from adaptdl.torch.scaling_rules import ScalingRuleBase
+
+class AdaptiveDLContext(object):
+    """
+    This class provides context tool to get AdaptDL-suggest parameters,
+    such as batch_size, accum_steps and lr_scale.
+    """
+
+    def __init__(self, batch_size):
+        self._elastic = adaptdl.torch.data.AdaptiveDataLoaderHelper(batch_size)
```
Comment (aurickq) [inline adaptdl/adaptdl/torch/context.py:31]: Should make sure the `AdaptiveDLContext` class does not depend on the `AdaptiveDataLoaderHelper` class (or anything related to PyTorch).

File: adaptdl/adaptdl/torch/context.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                adaptdl.env.num_nodes(), adaptdl.env.num_replicas(),
+                self._elastic.current_local_bsz, self._elastic.accumulation_steps)
+            # use only if speedup is significant
+            speedup = suggest_goodput / max(current_goodput, 1e-8)
+            if speedup > self._speedup_threshold:
+                self._elastic._state.current_local_bsz = atomic_bsz
+                self._elastic._state.accumulation_steps = accum_steps
+        self._elastic._state.current_local_bsz, self._elastic._state.accumulation_steps = \
+            adaptdl.collective.broadcast((self._elastic._state.current_local_bsz,
+                                          self._elastic._state.accumulation_steps))
```
Comment (aurickq) [inline adaptdl/adaptdl/torch/context.py:89]: Shouldn't do any cross-replica synchronization from a call to the `AdaptiveDLContext` class. The synchronization should happen outside, from the callsite.

File: adaptdl/adaptdl/torch/context.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        return self._elastic.current_local_bsz, self._elastic._state.current_local_bsz, self._elastic._state.accumulation_steps
+
+    @property
+    def training(self):
+        return self._elastic.training
+
+    def to_tensorboard(self, writer, global_step, tag_prefix=""):
+        self._elastic.to_tensorboard(writer, global_step, tag_prefix)
```
Comment (aurickq) [inline adaptdl/adaptdl/torch/context.py:97]: Tensorboard isn't core AdaptDL logic so shouldn't be included in this class.

File: adaptdl/adaptdl/torch/context.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+# limitations under the License.
+
+import collections
+import math
+import adaptdl.checkpoint
+import adaptdl.collective
+import adaptdl.env
+from adaptdl.torch._metrics import get_goodput_fn
+import adaptdl.torch.data
+from adaptdl.torch.scaling_rules import ScalingRuleBase
```
Comment (aurickq) [inline adaptdl/adaptdl/torch/context.py:22]: Should make sure this module doesn't depend on anything from PyTorch

File: adaptdl/adaptdl/torch/context.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+class AdaptiveDLContext(object):
```
Comment (aurickq) [inline adaptdl/adaptdl/torch/context.py:24]: Maybe rename it to just `Context` so usage of it can be simply `adaptdl.Context`

**Review by aurickq — CHANGES_REQUESTED** [review]

**Comment by Xuezhi-Liang** [comment]
close via a new PR #123

**Pull request closed without merging**